lib/pharos/salt.rb

```diff
@@ -16,54 +15,113 @@ def initialize
     # Returns two values:
     #   - The response object.
     #   - A hash containing the parsed JSON response.
-    def call(tgt, fun, arg = nil)
-      login! unless @token
-
+    def call(tgt:, fun:, arg: nil)
       hsh = { tgt: tgt, fun: fun, client: "local" }
       hsh[:arg] = arg if arg
 
-      res = perform_request("/", "post", hsh)
+      res = perform_request(endpoint: "/", method: "post", data: hsh)
       [res, JSON.parse(res.body)]
     end
 
     # Returns the minions as discovered by salt.
-    def minions(mid = nil)
-      login! unless @token
-
-      id  = mid ? "/#{mid}" : ""
-      res = perform_request("/minions#{id}", "get")
+    def minions
+      res = perform_request(endpoint: "/minions", method: "get")
       JSON.parse(res.body)["return"].first
     end
 
+    # Return information for the minion linked to this instance.
+    def minion(mid:)
+      res = perform_request(endpoint: "/minions/#{mid}", method: "get")
+      JSON.parse(res.body)["return"].first[mid]
+    end
+
+    class << self
+      # Call the orchestration
+      def orchestrate
+        perform_run_request(data: { client: "runner_async",
+                                    fun: "state.orchestrate",
+                                    kwargs: { mods: "orch.kubernetes" } })
+      end
+    end
+
+    def has_roles?(mid:)
```

Comment:
Yes, at least for now. If you have an already provisioned machine with roles assigned, you don't want to assign more roles to it, and that needs to be checked from the outside, or maybe ignore the assignment on `assign_role` method.